faunadb/strconv.go

```diff
@@ -0,0 +1,898 @@
+		return printFn("Scoped"+name, arg1, scope)
+	}
+	return printFn(name, arg1)
+}
+
+func printKeyScopedFn(name string, expr Expr) string {
+	if expr != Null() {
+		return printFn("Scoped"+name, expr)
+	}
+	return printNoArgsFn(name)
+
+}
+
+func printNoArgsFn(name string) string {
+	return _fnCall(name, noArgs, nil, nil)
+}
+
+func _fnCall(name string, argType int, mapping unescapedObj, args ...Expr) string {
+	pattern := "%s(%s)"
+	res := ""
+	switch argType {
+	case noArgs:
+		res = fmt.Sprintf(pattern, name, "")
+	case basicArgs:
+		strArgs := make([]string, len(args))
+		for idx, v := range args {
+			strArgs[idx] = v.String()
+		}
+		res = fmt.Sprintf(pattern, name, strings.Join(strArgs, ", "))
+	case varArgs:
+		if len(args) == 1 && reflect.ValueOf(args).Type().Implements(arrType) {
+			nestedArgs := reflect.ValueOf(args).Interface().(unescapedArr)
+			strArgs := make([]string, len(args))
+			for idx, v := range nestedArgs {
+				strArgs[idx] = v.String()
+			}
+			res = fmt.Sprintf(pattern, name, strings.Join(strArgs, ", "))
+		} else {
+			res = _fnCall(name, basicArgs, mapping, args...)
+		}
+	case optArgs:
+		if mapping == nil || len(mapping) == 0 {
+			panic(fmt.Sprintf("%s.String() accepts optional parameters, should have a non-zero mapping", name))
+		}
+		strArgs := []string{}
+		for _, v := range args {
+			strArgs = append(strArgs, v.String())
+		}
+		for k, v := range mapping {
+			if v != nil {
+				strArgs = append(strArgs, fmt.Sprintf(k, v))
+			}
+		}
+		res = fmt.Sprintf(pattern, name, strings.Join(strArgs, ", "))
+	}
+	return res
+}
+
+func printVarArgFn(name string, args ...Expr) string {
+	return _fnCall(name, varArgs, nil, args...)
+}
+
+func printFnWithOpts(name string, mapping unescapedObj, args ...Expr) string {
+	return _fnCall(name, optArgs, mapping, args...)
+}
+
+func (fn legacyRefFn) String() string {
+	return printFn("Ref", fn.Ref)
+}
+
+func (fn refFn) String() string {
+	return printFn("RefCollection", fn.Ref, fn.ID)
+}
+
+func (fn abortFn) String() string {
+	return printFn("Abort", fn.Abort)
+}
+
+func (fn doFn) String() string {
+	return printFn("Do", fn.Do)
+}
+
+func (fn ifFn) String() string {
+	return printFn("If", fn.If, fn.Then, fn.Else)
+}
+
+func (fn lambdaFn) String() string {
+	return printFn("Lambda", fn.Lambda, fn.Expression)
+}
+
+func (fn atFn) String() string {
+	return printFn("At", fn.At, fn.Expression)
+}
+
+func (fn letFn) String() string {
+	bindings := []string{}
+	tpe := reflect.TypeOf(fn.Let)
+	if tpe == reflect.TypeOf(unescapedArr{}) {
+		arr := fn.Let.(unescapedArr)
+		for _, elem := range arr {
+			m := elem.(unescapedObj)
+			for k, v := range m {
+				bindings = append(bindings, fmt.Sprintf("Bind(%s, %s)", k, v))
+			}
+		}
+	} else {
+		panic(fmt.Sprintf("Let bindings should be an unescapedArr, %s instead", tpe))
+	}
+	return fmt.Sprintf("Let().%s.In(%s)", strings.Join(bindings, "."), fn.In)
+}
+
+func (fn varFn) String() string {
+	return printFn("Var", fn.Var)
+}
```

Comment:
I think this is a place where we could use dynamic derivation on the function structs since it doesn't comprise runtime performance. For instance, the default implementation of `String()` for `fnApply` could simply reflect on the struct fields and generate the correct string representation. That'd require fields to be declared in the right order, though. For instance:

```
type subStringFn struct {
	fnApply
	SubString Expr `json:"substring"`
	Start     Expr `json:"start"`
	Length    Expr `json:"length,omitempty"`
}
```

Could be automatically derived into:

```
SubString(0, Length(10))
```

Tip: we can safely implement more tags (`faunarepr:"something"`) if we need more information about the declared fields.

Reply:
I'm generally against using reflect for this.

For one, coupling ordering of fields to the generated function signature doesn't seem ideal.

Dynamically deriving on the function structs would result in several reflect function calls which could be achieved using just a string builder(reflect is only used for varArg fns). IMO the overhead of retrieving a reflect.Type from a parameter, checking if its a struct and then iterating each field(and potentially looking up a tag) is not worth the reduction in code size.

Reply:
I'm not opposed to having static static functions. It's just tedious to maintain overtime. String representation is a non-performance critical util so, reflection isn't a concern here. 

Sorting struct fields based on the function spec order helps legibility. Now they match the test cases, in which their shape was dictated by the previous implementation that relied on golang's map key order. That' a bit counter intuitive.